File: server/app/schemas/closures.py
from marshmallow import Schema, fields, EXCLUDE


class ClosureSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    id            = fields.Int(dump_only=True)
    restaurant_id = fields.Int()
    start_date    = fields.Str(description="Start date (YYYY-MM-DD)")
    end_date      = fields.Str(description="End date (YYYY-MM-DD), equals start_date for single-day closure")
    reason        = fields.Str(allow_none=True, description="Short reason (e.g. 'Vacances scolaires')")
    description   = fields.Str(allow_none=True, description="Optional longer description")
    is_active     = fields.Bool()
    is_current    = fields.Bool(dump_only=True, description="True if today is within the closure range")
    notified_7d   = fields.Bool(dump_only=True)
    notified_1d   = fields.Bool(dump_only=True)
    created_at    = fields.Str(dump_only=True, allow_none=True)
    updated_at    = fields.Str(dump_only=True, allow_none=True)


class ClosureCreateSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    start_date    = fields.Str(required=True, description="Start date (YYYY-MM-DD)")
    end_date      = fields.Str(required=True, description="End date (YYYY-MM-DD)")
    reason        = fields.Str(allow_none=True)
    description   = fields.Str(allow_none=True)
    restaurant_id = fields.Int()


class ClosureUpdateSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    start_date  = fields.Str(description="Start date (YYYY-MM-DD)")
    end_date    = fields.Str(description="End date (YYYY-MM-DD)")
    reason      = fields.Str(allow_none=True)
    description = fields.Str(allow_none=True)
    is_active   = fields.Bool()


class PublicClosuresResponseSchema(Schema):
    class Meta:
        unknown = EXCLUDE
    current_closure   = fields.Dict(allow_none=True)
    upcoming_closures = fields.List(fields.Dict())
    closures          = fields.List(fields.Dict(), description="All closures (editor view)")

File: client/src/lib/api/closures.ts
import { api, publicAxios, PUBLIC_API_TIMEOUT_MS } from './client';

export interface ExceptionalClosure {
  id: number;
  restaurant_id: number;
  start_date: string; // YYYY-MM-DD
  end_date: string; // YYYY-MM-DD — equals start_date for single-day closure
  reason?: string;
  description?: string;
  is_active: boolean;
  is_current: boolean;
  notified_7d: boolean;
  notified_1d: boolean;
  created_at?: string;
  updated_at?: string;
}

export const closuresApi = {
  list: async (upcoming = false, restaurantId?: number, includeInactive = false) => {
    const params: Record<string, string | number | boolean> = {
      upcoming: String(upcoming),
      include_inactive: String(includeInactive),
    };
    if (restaurantId) params.restaurant_id = restaurantId;
    const response = await api.get('/closures', { params });
    return response.data.closures as ExceptionalClosure[];
  },

  create: async (closure: Partial<ExceptionalClosure>) => {
    const response = await api.post('/closures', closure);
    return response.data.closure as ExceptionalClosure;
  },

  update: async (id: number, closure: Partial<ExceptionalClosure>) => {
    const response = await api.put(`/closures/${id}`, closure);
    return response.data.closure as ExceptionalClosure;
  },

  delete: async (id: number) => {
    await api.delete(`/closures/${id}`);
  },

  getPublic: async (restaurantSlug: string) => {
    const response = await publicAxios.get(`/public/${restaurantSlug}/closures`, {
      timeout: PUBLIC_API_TIMEOUT_MS,
    });
    return response.data as {
      current_closure: ExceptionalClosure | null;
      upcoming_closures: ExceptionalClosure[];
      closures: ExceptionalClosure[];
    };
  },
};
